File: frontend/scratch/optimizer.py
import os
import re

dirs = [
    os.path.join(os.path.dirname(__file__), '../src/pages/Inicio'),
    os.path.join(os.path.dirname(__file__), '../src/components')
]

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    original_content = content

    def button_repl(m):
        before = m.group(1)
        quote1 = m.group(2)
        classes = m.group(3)
        quote2 = m.group(4)
        if 'transform-gpu' not in classes:
            return f'<button{before}className={quote1}{classes} transform-gpu{quote2}'
        return m.group(0)

    content = re.sub(r'<button([^>]*)className=(["\'{`])([^"\'{`]*)(["\'{`])', button_repl, content)

    def class_repl(m):
        classes = m.group(1)
        new_classes = classes

        if ('hover:scale' in new_classes or 'translate-x' in new_classes or 'translate-y' in new_classes or '-translate' in new_classes) and 'transform-gpu' not in new_classes:
            new_classes += ' transform-gpu'
            
        if 'transition-all' in new_classes:
            if 'scale' in new_classes or 'translate' in new_classes:
                replacement = 'transition-transform duration-200 ease-out'
            elif 'opacity-0' in new_classes and 'opacity-100' in new_classes:
                replacement = 'transition-opacity duration-200 ease-out'
            elif 'h-full' in new_classes and 'rounded-full' in new_classes and ('bg-primary' in new_classes or 'bg-emerald' in new_classes):
                replacement = 'transition-[width] duration-500 ease-out'
            elif 'max-w-0' in new_classes or 'max-w-xs' in new_classes:
                replacement = 'transition-all duration-300 ease-out'
            else:
                replacement = 'transition-colors duration-150'
                
            new_classes = new_classes.replace('transition-all', replacement)
            
            if 'duration-' in replacement:
                rep_duration = re.search(r'duration-\d+', replacement).group(0)
                new_classes = re.sub(r'duration-\d+', lambda match: match.group(0) if match.group(0) == rep_duration else '', new_classes)
            if 'ease-' in replacement:
                rep_ease = re.search(r'ease-[a-z-]+', replacement).group(0)
                new_classes = re.sub(r'ease-(in|out|in-out)', lambda match: match.group(0) if match.group(0) == rep_ease else '', new_classes)

        new_classes = re.sub(r'\s+', ' ', new_classes).strip()
        return m.group(0).replace(classes, new_classes)

    content = re.sub(r'className=(?:\{`|["\'])(.*?)(?:`\}|["\'])', class_repl, content)

    if content != original_content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        print('Updated:', filepath)

for d in dirs:
    for root, _, files in os.walk(d):
        for file in files:
            if file.endswith('.tsx') or file.endswith('.jsx') or file.endswith('.ts') or file.endswith('.js'):
                process_file(os.path.join(root, file))

File: frontend/scratch/optimizer.js
const fs = require('fs');
const path = require('path');

const dirs = [
  path.join(__dirname, '../../src/pages/Inicio'),
  path.join(__dirname, '../../src/components')
];

function processFile(filePath) {
  let content = fs.readFileSync(filePath, 'utf-8');
  let originalContent = content;

  // 1. Add transform-gpu to <button>
  content = content.replace(/<button([^>]*)className=(["'{`])([^"'{`]*)(["'{`])/g, (match, before, quote1, classes, quote2) => {
    if (!classes.includes('transform-gpu')) {
      return `<button${before}className=${quote1}${classes} transform-gpu${quote2}`;
    }
    return match;
  });

  // 2. Add transform-gpu to elements with hover:scale-* or hover:-translate-* or translate-x-*
  // A bit complex with template literals, let's just do a global replace for className="..." or className={...}
  const classRegex = /className=(?:\{`|["'])(.*?)(?:`\}|["'])/g;
  
  content = content.replace(classRegex, (match, classes) => {
    let newClasses = classes;
    
    // Add transform-gpu if it has hover:scale-* or translate-x-* or translate-y-*
    if ((newClasses.includes('hover:scale') || newClasses.includes('translate-x') || newClasses.includes('translate-y') || newClasses.includes('-translate')) && !newClasses.includes('transform-gpu')) {
      newClasses += ' transform-gpu';
    }

    // Replace transition-all based on context
    if (newClasses.includes('transition-all')) {
      let replacement = '';
      if (newClasses.includes('scale') || newClasses.includes('translate')) {
        replacement = 'transition-transform duration-200 ease-out';
      } else if (newClasses.includes('opacity-0') && newClasses.includes('opacity-100')) {
        replacement = 'transition-opacity duration-200 ease-out';
      } else if (newClasses.includes('w-') || newClasses.includes('h-full')) {
        // Maybe progress bar, but let's check for specific files or class combinations
        if (newClasses.includes('h-full') && newClasses.includes('rounded-full') && (newClasses.includes('bg-primary') || newClasses.includes('bg-emerald'))) {
          replacement = 'transition-[width] duration-500 ease-out'; // Specific to TopKpiCards progress bars
        } else if (newClasses.includes('max-w-0') || newClasses.includes('max-w-xs')) {
           replacement = 'transition-all duration-300 ease-out'; // keep as is for max-w since transition-[max-width] might be needed, or just transition-all
           replacement = 'transition-[max-width,opacity] duration-300 ease-out';
        } else {
           replacement = 'transition-colors duration-150';
        }
      } else {
        replacement = 'transition-colors duration-150';
      }
      
      // Cleanup existing durations if we added new ones
      newClasses = newClasses.replace('transition-all', replacement);
      // Remove old durations if we added new ones from the replacement, to avoid duplicates
      if (replacement.includes('duration-')) {
         newClasses = newClasses.replace(/duration-\d+/g, (match) => {
             return match === replacement.match(/duration-\d+/)[0] ? match : '';
         });
      }
      if (replacement.includes('ease-')) {
         newClasses = newClasses.replace(/ease-(in|out|in-out)/g, (match) => {
             return match === replacement.match(/ease-[a-z-]+/)[0] ? match : '';
         });
      }
    }

    // Clean up multiple spaces
    newClasses = newClasses.replace(/\s+/g, ' ').trim();
    
    // Replace the classes in the match string
    return match.replace(classes, newClasses);
  });

  if (content !== originalContent) {
    fs.writeFileSync(filePath, content);
    console.log('Updated:', filePath);
  }
}

function walk(dir) {
  const list = fs.readdirSync(dir);
  for (const file of list) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
      walk(filePath);
    } else if (filePath.endsWith('.tsx') || filePath.endsWith('.jsx') || filePath.endsWith('.ts') || filePath.endsWith('.js')) {
      processFile(filePath);
    }
  }
}

dirs.forEach(walk);
